modules/bmc/redfish.rb

```diff
@@ -0,0 +1,232 @@
+require 'redfish_client'
+require 'bmc/base'
+require 'bmc/redfish/dell'
+require 'bmc/redfish/hpe'
+
+module Proxy
+  module BMC
+    class Redfish < Base
+      include Proxy::Log
+      include Proxy::Util
+
+      def initialize(args)
+        @redfish_verify_ssl = Proxy::BMC::Plugin.settings.bmc_redfish_verify_ssl
+        # enforce SSL verification as a default, if the value isn't set in the config
+        # and bail if it's set to something nonsensical
+        @redfish_verify_ssl = true if @redfish_verify_ssl.nil?
+        raise ::Proxy::Error::ConfigurationError.new("bmc_redfish_verify_ssl must be boolean (true/false)") unless [true, false].include?(@redfish_verify_ssl)
+        super
+        load_vendor_overrides
+      end
+
+      def connect(args = { })
+        connection = RedfishClient.new("https://#{args[:host]}/", verify: @redfish_verify_ssl)
+        connection.login(args[:username], args[:password])
+        connection
+      end
+
+      def load_vendor_overrides
+        mod = case manufacturer
+              when 'Dell Inc.' then 'RedfishVendorOverridesDellInc'
+              when 'HPE'       then 'RedfishVendorOverridesHPE'
+              else
+                logger.debug "No #{manufacturer} specific overrides available - using generic Redfish calls"
+                return
+              end
+        logger.debug "Extending Redfish with vendor overrides for #{manufacturer}"
+        extend Kernel.const_get(mod.to_sym)
```

Comment:
If it's done once during initialization, the provider can only ever work with a single vendor's redfish implementation, which defeats the point. We have systems from four different vendors in production, and some BMCs behave differently even within a single vendor, owing to function changes between versions. It has to happen at runtime, based on the specific Redfish implementation that BMC proxy instance is communicating with.

Unless you're advocating for an arrangement where there are case statements in every method to support every possible vendor? I don't consider that an improvement in any way, but if that's the will of the maintainers, that's what I'll do. Or... something else I don't understand yet?

Reply:
Actually.. is it the extend specifically? I just recently learned I don't quiiite understand extend the way I thought I did, so there may be side effects I don't understand. But my intent was to allow some methods to be overridden at runtime based on data coming back from the specific BMC, and (secondarily) to allow the use of super to simplify those overrides (since in many cases it'll just be static data that needs changing, either values passed to an endpoint, or the location of an endpoint; the interaction with an endpoint otherwise being the same).

Reply:
I did an experiment to check some of the assumptions I'd had about extend's outward behavior and... although I still don't quite understand the way it works, it does at least behave the way I'd assumed - extending gets the module's methods available in that single instance. I had read (and I am probably butchering this, since I don't yet understand it, etc.) that extending worked on the class's singleton object -- that had me worried that perhaps the methods were becoming available to all instances, and the Class itself would be dragged back and forth between one vendor or another's overrides. Anyway, that doesn't seem to be the case. For whatever that's worth.

Reply:
It took me a bit to wrap my head around what's going on and I must say I'm impressed, this is genius. A bit on the crazy side, but genius nevertheless.

I would still prefer having it done the boring way with some kind of adapters or something, but this should work.

Reply:
well, I just noticed I missed a trick changing it from the dynamic lookup - no need to assign a string to mod, then convert it to symbol. I could have just assigned the symbol directly to mod.

I wouldn't mind learning what you mean by "the boring way with some kind of adapters or something", if you have the inclination to explain it. Just for my own growth.

Reply:
Something along the lines of this https://github.com/theforeman/smart-proxy/commit/655a8c711c005ecfd024f0b39b894420f09a78db . It is a bit more explicit than what you have here. And also untested :)

Reply:
I see. The cost of doing it that way is a lot of extra round trips to the BMC -- one must connect, authenticate, and query the manufacturer, before creating the new object which must connect, authenticate, etc. (Unless, I suppose, one passes a reference to the original session to the new object.) 

What are the benefits?

Reply:
The connection is indeed passed [here](https://github.com/theforeman/smart-proxy/commit/9889bc5f640d273a20c1c30f74f788a1fc5ac9aa#diff-3b223db7ff1577c5ae9efd3ac7a328c6R32) so the it gets opened only once.

I'd say the benefits are that the code is more explicit about what is going on and one doesn't need to know advanced ruby stuff like that extend when called within an instance extends the instance's eigenclass.

Reply:
I missed that. Actually I will need to look closer at this because you've made redfish_adapter a class method, which then apparently has the somewhat perverse side effect of returning an instance of itself. Believe it or not this is more confusing to me! But I guess it could be thought of as a somewhat pathological angle on new(), which makes it seem less threatening. Is this what Java and Python people refer to as a "factory" pattern?

Reply:
I guess what I'm driving at, is... I think there's an element in both of these directions that effectively boils down to, "this is what's familiar to me." I agree that there's value in doing things that are understood, and in that case your preference as a maintainer I think legitimately carries more weight. To counter that, all I have to say is something like... I think it's fair to say Foreman is pretty solidly committed to ruby. Why avoid taking advantage of legit features of the language in favor of something that feels to me more like a Java pattern? It's admittedly not much of an argument. If it's your will that I change it, I will; however reluctantly.

Reply:
> Is this what Java and Python people refer to as a "factory" pattern?

Most likely yes, but I must admit I'm a bit rusty with design patterns :)

I'd be fine with your way if you just added a comment briefly explaining why it works the way it does so it is not so surprising. Not sure how others look at this though

Reply:
Let's do this; it'd be simple for me to add a comment. So let me do that (and fix the assignment thing I mentioned) while you guys come to a consensus on keeping it or not. If I may ask -- could you describe what you find surprising about the use of extend? It would help me write a more useful comment.

Reply:
Took a stab at it; focused on concrete effects rather than anything philosophical. Let me know if it misses the mark?

Reply:
I am reading this right now, sorry for the delay. Let me review what I thought when I made that short comment :-)

Reply:
Allright, my first quick impression is this code smells and there's the traditional way of workarounding it, for example the one that Adam described. Since you insist on the Ruby way, and it's totally fine here this indeed is a Ruby project, I had to drill down the codebase to confirm some things. One of my biggest concern was - what if we keep the Redfish instance in memory? Then we will be extending the singleton class on every request until the call chain is long AF practically slowing down every processing by a bit. That's not the case, we create new instance on every request, at least this is what I expect from Sinatra/Rack contrast. Note if this is fine today, this might change in the future creating terrible bug that will cost a lot of money to track it down (customers complaining that the service slows down over time, memory consumption etc).

My second concern is performance, this was just an impression without raw numbers but quick google finds that creating an object and then extending [it is nearly 4 times as slow](https://stackoverflow.com/questions/8616275/how-costly-is-extend-in-ruby) as only creating the object. Probably not the big issue for Redfish where this request will spend 99% in waiting for I/O, but for me it's a flag raised.

And finally, I am ex-Java guy - Hello! :-) - and I still find the traditional design patterns easier to understand for an average software engineer. You need to understand who is gonna maintain the code - we are most commonly people with some traditional programming background (java, c/c++, python etc) than native-rubyists. It's not the language what defines how things should be written, it's the project and people around it.

Having said that, I was ultra slow responding to your PR, this is a feature we want and if I am not mistaken there are no known bugs other than the minor performance concern. If you can confirm @adamruzicka that this extension approach is correct, let's merge this.

Reply:
:+1: from me